x/pairing/types/message_stake_provider.go

```diff
@@ -55,5 +58,47 @@ func (msg *MsgStakeProvider) ValidateBasic() error {
 	if len(msg.Moniker) > MAX_LEN_MONIKER {
 		return sdkerrors.Wrapf(MonikerTooLongError, "invalid moniker (%s)", msg.Moniker)
 	}
+
+	err = ValidateGeoFields(msg.Endpoints, msg.Geolocation)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// verify that the geolocation arg is a union of the endpoints' geolocation
+func ValidateGeoFields(endp []epochstoragetypes.Endpoint, geo uint64) error {
+	geoSeen := map[string]struct{}{}
+	var endpointsGeoStr string
+	for _, endp := range endp {
+		geoStr := planstypes.Geolocation_name[int32(endp.Geolocation)]
+		_, ok := geoSeen[geoStr]
+		if !ok {
+			geoSeen[geoStr] = struct{}{}
+			endpointsGeoStr += geoStr + ","
+		}
+	}
+
+	endpointsGeoStr = strings.TrimSuffix(endpointsGeoStr, ",")
+	geoEnums, geoStr := ExtractGeolocations(geo)
+	if len(geoEnums) != len(geoSeen) {
+		return sdkerrors.Wrapf(GeolocationNotMatchWithEndpointsError,
+			"invalid geolocation (endpoints combined geolocation: {%s}, provider geolocation: {%s})", endpointsGeoStr, geoStr)
+	}
+	for _, geoE := range geoEnums {
+		_, ok := geoSeen[geoE.String()]
+		if !ok {
+			if geoE != planstypes.Geolocation_GL {
+				return sdkerrors.Wrapf(GeolocationNotMatchWithEndpointsError,
+					"invalid geolocation (endpoints combined geolocation: {%s}, provider geolocation: {%s})", endpointsGeoStr, geoStr)
+			} else if len(geoSeen) != len(planstypes.Geolocation_name)-2 {
+				// handle the global case (should see all geos minus 2 global geos)
+				return sdkerrors.Wrapf(GeolocationNotMatchWithEndpointsError,
+					"invalid global geolocation (endpoints combined geolocation: {%s}, provider geolocation: {%s})", endpointsGeoStr, geoStr)
+			}
+		}
+	}
+
 	return nil
```

Comment:
deleted this function since we make the exact same check in your modified `HandleEndpointsAndGeolocationArgs`. see https://github.com/lavanet/lava/pull/577/commits/b5b9157e060150e9f6dd6d32b60743b382857ebd